File: src/main/java/lechuck/intellij/vars/VariablesTable.java
package lechuck.intellij.vars;

import com.intellij.execution.ExecutionBundle;
import com.intellij.execution.util.ListTableWithButtons;
import com.intellij.execution.util.StringWithNewLinesCellEditor;
import com.intellij.icons.AllIcons;
import com.intellij.ide.CopyProvider;
import com.intellij.ide.PasteProvider;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.ide.CopyPasteManager;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.AnActionButton;
import com.intellij.ui.table.TableView;
import com.intellij.util.containers.ContainerUtil;
import com.intellij.util.ui.ListTableModel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.table.TableCellEditor;
import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.StringSelection;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class VariablesTable extends ListTableWithButtons<Variable> {
    private CopyPasteProviderPanel myPanel;
    private boolean myPasteEnabled = false;

    public VariablesTable() {
        getTableView().getEmptyText().setText(ExecutionBundle.message("empty.text.no.variables"));
        AnAction copyAction = ActionManager.getInstance().getAction(IdeActions.ACTION_COPY);
        if (copyAction != null) {
            copyAction.registerCustomShortcutSet(copyAction.getShortcutSet(), getTableView()); // no need to add in popup menu
        }
        AnAction pasteAction = ActionManager.getInstance().getAction(IdeActions.ACTION_PASTE);
        if (pasteAction != null) {
            pasteAction.registerCustomShortcutSet(pasteAction.getShortcutSet(), getTableView()); // no need to add in popup menu
        }
    }

    public void setPasteActionEnabled(boolean enabled) {
        myPasteEnabled = enabled;
    }

    @Override
    protected ListTableModel createListModel() {
        return new ListTableModel(new NameColumnInfo(), new ValueColumnInfo());
    }

    public void editVariableName(final Variable variable) {
        ApplicationManager.getApplication().invokeLater(() -> {
            final Variable actualVar = ContainerUtil.find(getElements(),
                    item -> StringUtil.equals(variable.getName(), item.getName()));
            if (actualVar == null) {
                return;
            }

            setSelection(actualVar);
            editSelection(0);
        });
    }

    public List<Variable> getVariables() {
        return getElements();
    }

    @Override
    public JComponent getComponent() {
        if (myPanel == null) {
            myPanel = new CopyPasteProviderPanel(super.getComponent());
        }
        return myPanel;
    }

    @Override
    protected Variable createElement() {
        return new Variable("", "");
    }

    @Override
    protected boolean isEmpty(Variable element) {
        return element.getName().isEmpty() && element.getValue().isEmpty();
    }


    @Override
    protected Variable cloneElement(Variable variable) {
        return variable.clone();
    }

    @Override
    protected boolean canDeleteElement(Variable selection) {
        return true;
    }

    protected class NameColumnInfo extends ElementsColumnInfoBase<Variable> {
        public NameColumnInfo() {
            super("Name");
        }

        @Override
        public String valueOf(Variable variable) {
            return variable.getName();
        }

        @Override
        public boolean isCellEditable(Variable variable) {
            return true;
        }

        @Override
        public void setValue(Variable variable, String s) {
            if (s.equals(valueOf(variable))) {
                return;
            }
            variable.setName(s);
            setModified();
        }

        @Override
        protected String getDescription(Variable variable) {
            return variable.getDescription();
        }

        @NotNull
        @Override
        public TableCellEditor getEditor(Variable variable) {
            return new DefaultCellEditor(new JTextField());
        }
    }

    protected class ValueColumnInfo extends ElementsColumnInfoBase<Variable> {
        public ValueColumnInfo() {
            super("Value");
        }

        @Override
        public String valueOf(Variable variable) {
            return variable.getValue();
        }

        @Override
        public boolean isCellEditable(Variable variable) {
            return true;
        }

        @Override
        public void setValue(Variable variable, String s) {
            if (s.equals(valueOf(variable))) {
                return;
            }
            variable.setValue(s);
            setModified();
        }

        @Nullable
        @Override
        protected String getDescription(Variable variable) {
            return variable.getDescription();
        }

        @NotNull
        @Override
        public TableCellEditor getEditor(Variable variable) {
            return new StringWithNewLinesCellEditor();
        }
    }

    private final class CopyPasteProviderPanel extends JPanel implements DataProvider, CopyProvider, PasteProvider {
        private CopyPasteProviderPanel(JComponent component) {
            super(new GridLayout(1, 1));
            add(component);
        }

        @Nullable
        @Override
        public Object getData(@NotNull String dataId) {
            if (PlatformDataKeys.COPY_PROVIDER.is(dataId) || PlatformDataKeys.PASTE_PROVIDER.is(dataId)) {
                return this;
            }
            return null;
        }

        @Override
        public void performCopy(@NotNull DataContext dataContext) {
            TableView<Variable> view = getTableView();
            if (view.isEditing()) {
                int row = view.getEditingRow();
                int column = view.getEditingColumn();
                if (row < 0 || column < 0) {
                    row = view.getSelectedRow();
                    column = view.getSelectedColumn();
                }
                if (row >= 0 && column >= 0) {
                    JTextField textField = (JTextField) ((DefaultCellEditor) view.getCellEditor()).getComponent();
                    CopyPasteManager.getInstance().setContents(new StringSelection(textField.getSelectedText()));
                }
                return;
            }
            stopEditing();
            StringBuilder sb = new StringBuilder();
            List<Variable> variables = getSelection();
            for (Variable variable : variables) {
                if (isEmpty(variable)) continue;
                if (sb.length() > 0) sb.append(';');
                sb.append(StringUtil.escapeChars(variable.getName(), '=', ';')).append('=')
                        .append(StringUtil.escapeChars(variable.getValue(), '=', ';'));
            }
            CopyPasteManager.getInstance().setContents(new StringSelection(sb.toString()));
        }


        @Override
        public boolean isCopyEnabled(@NotNull DataContext dataContext) {
            return !getSelection().isEmpty();
        }

        @Override
        public boolean isCopyVisible(@NotNull DataContext dataContext) {
            return isCopyEnabled(dataContext);
        }

        @Override
        public void performPaste(@NotNull DataContext dataContext) {
            String content = CopyPasteManager.getInstance().getContents(DataFlavor.stringFlavor);
            if (StringUtil.isEmpty(content)) {
                return;
            }
            Map<String, String> map = parseVarssFromText(content);
            TableView<Variable> view = getTableView();
            if ((view.isEditing() || map.isEmpty())) {
                int row = view.getEditingRow();
                int column = view.getEditingColumn();
                if (row < 0 || column < 0) {
                    row = view.getSelectedRow();
                    column = view.getSelectedColumn();
                }
                if (row >= 0 && column >= 0) {
                    TableCellEditor editor = view.getCellEditor();
                    if (editor != null) {
                        Component component = ((DefaultCellEditor) editor).getComponent();
                        if (component instanceof JTextField) {
                            ((JTextField) component).paste();
                        }
                    }
                }
                return;
            }
            stopEditing();
            List<Variable> parsed = new ArrayList<>();
            for (Map.Entry<String, String> entry : map.entrySet()) {
                parsed.add(new Variable(entry.getKey(), entry.getValue()));
            }
            List<Variable> variables = new ArrayList<>(getVariables());
            variables.addAll(parsed);
            variables = ContainerUtil.filter(variables, variable -> !StringUtil.isEmpty(variable.getName()) ||
                    !StringUtil.isEmpty(variable.getValue()));
            setValues(variables);
        }

        @Override
        public boolean isPastePossible(@NotNull DataContext dataContext) {
            return myPasteEnabled;
        }

        @Override
        public boolean isPasteEnabled(@NotNull DataContext dataContext) {
            return myPasteEnabled;
        }
    }

    @Override
    protected AnActionButton @NotNull [] createExtraActions() {
        AnActionButton copyButton = new AnActionButton("Copy", AllIcons.Actions.Copy) {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e) {
                myPanel.performCopy(e.getDataContext());
            }

            @Override
            public boolean isEnabled() {
                return myPanel.isCopyEnabled(DataContext.EMPTY_CONTEXT);
            }
        };
        AnActionButton pasteButton = new AnActionButton("Paste", AllIcons.Actions.MenuPaste) {
            @Override
            public void actionPerformed(@NotNull AnActionEvent e) {
                myPanel.performPaste(e.getDataContext());
            }

            @Override
            public boolean isEnabled() {
                return myPanel.isPasteEnabled(DataContext.EMPTY_CONTEXT);
            }

            @Override
            public boolean isVisible() {
                return myPanel.isPastePossible(DataContext.EMPTY_CONTEXT);
            }
        };
        return new AnActionButton[]{copyButton, pasteButton};
    }

    @NotNull
    public static Map<String, String> parseVarssFromText(String content) {
        Map<String, String> result = new LinkedHashMap<>();
        if (content != null && content.contains("=")) {
            boolean legacyFormat = content.contains("\n");
            List<String> pairs;
            if (legacyFormat) {
                pairs = StringUtil.split(content, "\n");
            } else {
                pairs = new ArrayList<>();
                int start = 0;
                int end;
                for (end = content.indexOf(";"); end < content.length(); end = content.indexOf(";", end + 1)) {
                    if (end == -1) {
                        pairs.add(content.substring(start).replace("\\;", ";"));
                        break;
                    }
                    if (end > 0 && (content.charAt(end - 1) != '\\')) {
                        pairs.add(content.substring(start, end).replace("\\;", ";"));
                        start = end + 1;
                    }
                }
            }
            for (String pair : pairs) {
                int pos = pair.indexOf('=');
                if (pos <= 0) continue;
                while (pos > 0 && pair.charAt(pos - 1) == '\\') {
                    pos = pair.indexOf('=', pos + 1);
                }
                pair = pair.replaceAll("[\\\\]", "\\\\\\\\");
                result.put(StringUtil.unescapeStringCharacters(pair.substring(0, pos)).trim(),
                        StringUtil.unescapeStringCharacters(pair.substring(pos + 1)));
            }
        }
        return result;
    }
}

File: src/main/kotlin/lechuck/intellij/vars/VariablesTable.kt
package lechuck.intellij.vars

import com.intellij.execution.ExecutionBundle
import com.intellij.execution.util.ListTableWithButtons
import com.intellij.icons.AllIcons
import com.intellij.ide.CopyProvider
import com.intellij.ide.PasteProvider
import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.ide.CopyPasteManager
import com.intellij.openapi.util.text.StringUtil
import com.intellij.ui.AnActionButton
import com.intellij.util.containers.ContainerUtil
import com.intellij.util.ui.ListTableModel
import java.awt.GridLayout
import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.StringSelection
import javax.swing.DefaultCellEditor
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.JTextField
import javax.swing.table.TableCellEditor

open class VariablesTable : ListTableWithButtons<Variable>() {
    private var myPanel: CopyPasteProviderPanel? = null
    private var myPasteEnabled = false

    init {
        tableView.emptyText.text = ExecutionBundle.message("empty.text.no.variables")
        val copyAction = ActionManager.getInstance().getAction(IdeActions.ACTION_COPY)
        copyAction?.registerCustomShortcutSet(
            copyAction.shortcutSet,
            tableView,
        ) // no need to add in popup menu
        val pasteAction = ActionManager.getInstance().getAction(IdeActions.ACTION_PASTE)
        pasteAction?.registerCustomShortcutSet(
            pasteAction.shortcutSet,
            tableView,
        ) // no need to add in popup menu
    }

    fun setPasteActionEnabled(enabled: Boolean) {
        myPasteEnabled = enabled
    }

    override fun createListModel(): ListTableModel<Variable> {
        return ListTableModel(NameColumnInfo(), ValueColumnInfo())
    }

    fun editVariableName(variable: Variable) {
        ApplicationManager.getApplication().invokeLater {
            val actualVar =
                ContainerUtil.find(getElements()) { item ->
                    StringUtil.equals(variable.name, item.name)
                }
            if (actualVar == null) {
                return@invokeLater
            }

            setSelection(actualVar)
            editSelection(0)
        }
    }

    fun getVariables(): List<Variable> {
        return getElements()
    }

    override fun getComponent(): JComponent {
        if (myPanel == null) {
            myPanel = CopyPasteProviderPanel(super.getComponent())
        }
        return myPanel as CopyPasteProviderPanel
    }

    override fun createElement(): Variable {
        return Variable("", "")
    }

    override fun isEmpty(element: Variable): Boolean {
        return element.name.isEmpty() && element.value.isEmpty()
    }

    override fun cloneElement(variable: Variable): Variable {
        return Variable(variable.name, variable.value)
    }

    override fun canDeleteElement(selection: Variable): Boolean {
        return true
    }

    protected open inner class NameColumnInfo : ElementsColumnInfoBase<Variable>("Name") {
        override fun valueOf(variable: Variable): String {
            return variable.name
        }

        override fun isCellEditable(variable: Variable): Boolean {
            return true
        }

        override fun setValue(variable: Variable, s: String) {
            if (s == valueOf(variable)) {
                return
            }
            variable.name = s
            setModified()
        }

        override fun getDescription(variable: Variable): String? {
            return variable.getDescription()
        }

        override fun getEditor(variable: Variable): TableCellEditor {
            return createCellEditor()
        }
    }

    protected open inner class ValueColumnInfo : ElementsColumnInfoBase<Variable>("Value") {
        override fun valueOf(variable: Variable): String {
            return variable.value
        }

        override fun isCellEditable(variable: Variable): Boolean {
            return true
        }

        override fun setValue(variable: Variable, s: String) {
            if (s == valueOf(variable)) {
                return
            }
            variable.value = s
            setModified()
        }

        override fun getDescription(variable: Variable): String? {
            return variable.getDescription()
        }

        override fun getEditor(variable: Variable): TableCellEditor {
            return createCellEditor()
        }
    }

    private inner class CopyPasteProviderPanel(component: JComponent) :
        JPanel(GridLayout(1, 1)), DataProvider, CopyProvider, PasteProvider {

        init {
            add(component)
        }

        override fun getData(dataId: String): Any? {
            return if (
                PlatformDataKeys.COPY_PROVIDER.`is`(dataId) ||
                    PlatformDataKeys.PASTE_PROVIDER.`is`(dataId)
            ) {
                this
            } else null
        }

        override fun performCopy(dataContext: DataContext) {
            val view = tableView
            if (view.isEditing) {
                var row = view.editingRow
                var column = view.editingColumn
                if (row < 0 || column < 0) {
                    row = view.selectedRow
                    column = view.selectedColumn
                }
                if (row >= 0 && column >= 0) {
                    val textField = (view.cellEditor as DefaultCellEditor).component as JTextField
                    CopyPasteManager.getInstance()
                        .setContents(StringSelection(textField.selectedText))
                }
                return
            }
            stopEditing()
            val text = stringifyForCopy(selection.filterNot { isEmpty(it) })
            CopyPasteManager.getInstance().setContents(StringSelection(text))
        }

        override fun isCopyEnabled(dataContext: DataContext): Boolean {
            return selection.isNotEmpty()
        }

        override fun isCopyVisible(dataContext: DataContext): Boolean {
            return isCopyEnabled(dataContext)
        }

        override fun performPaste(dataContext: DataContext) {
            val content =
                CopyPasteManager.getInstance().getContents<String>(DataFlavor.stringFlavor)
            if (StringUtil.isEmpty(content)) {
                return
            }
            val map = parseVarsFromText(content)
            val view = tableView
            if (view.isEditing || map.isEmpty()) {
                var row = view.editingRow
                var column = view.editingColumn
                if (row < 0 || column < 0) {
                    row = view.selectedRow
                    column = view.selectedColumn
                }
                if (row >= 0 && column >= 0) {
                    val editor = view.cellEditor
                    if (editor != null) {
                        val component = (editor as DefaultCellEditor).component
                        if (component is JTextField) {
                            component.paste()
                        }
                    }
                }
                return
            }
            stopEditing()
            val parsed = ArrayList<Variable>()
            for ((key, value) in map) {
                parsed.add(Variable(key, value))
            }
            var variables = ArrayList(getVariables())
            variables.addAll(parsed)
            variables =
                ArrayList(
                    ContainerUtil.filter(variables) { variable ->
                        !StringUtil.isEmpty(variable.name) || !StringUtil.isEmpty(variable.value)
                    }
                )
            setValues(variables)
        }

        override fun isPastePossible(dataContext: DataContext): Boolean {
            return myPasteEnabled
        }

        override fun isPasteEnabled(dataContext: DataContext): Boolean {
            return myPasteEnabled
        }
    }

    override fun createExtraToolbarActions(): Array<AnAction> {
        val copyButton =
            object : AnActionButton("Copy", AllIcons.Actions.Copy) {
                override fun actionPerformed(e: AnActionEvent) {
                    myPanel?.performCopy(e.dataContext)
                }

                override fun isEnabled(): Boolean {
                    return myPanel?.isCopyEnabled(DataContext.EMPTY_CONTEXT) == true
                }

                override fun getActionUpdateThread(): ActionUpdateThread {
                    return ActionUpdateThread.EDT
                }
            }
        val pasteButton =
            object : AnActionButton("Paste", AllIcons.Actions.MenuPaste) {
                override fun actionPerformed(e: AnActionEvent) {
                    myPanel?.performPaste(e.dataContext)
                }

                override fun isEnabled(): Boolean {
                    return myPanel?.isPasteEnabled(DataContext.EMPTY_CONTEXT) == true
                }

                override fun isVisible(): Boolean {
                    return myPanel?.isPastePossible(DataContext.EMPTY_CONTEXT) == true
                }

                override fun getActionUpdateThread(): ActionUpdateThread {
                    return ActionUpdateThread.EDT
                }
            }
        return arrayOf(copyButton, pasteButton)
    }

    companion object {
        /**
         * Characters a backslash may escape. A backslash before anything else is a literal
         * character, which is what lets a Windows path like `C:\temp\new` be typed as-is.
         *
         * The backslash itself must be in this set. Without it the encoding is not reversible, and
         * a value ending in a backslash would swallow the following variable.
         *
         * The cost is that text from elsewhere is read by these rules too, so a pasted `\\server`
         * arrives as `\server`. Reversibility for what this table writes is worth more.
         */
        private const val ESCAPABLE = "\\;="

        /** Both columns use a plain text field, whose document filters newlines. */
        internal fun createCellEditor(): TableCellEditor = DefaultCellEditor(JTextField())

        /**
         * Reads the `name=value;name=value` form written by [stringifyForCopy] and by
         * `VariablesTextFieldWithBrowseButton.stringifyVars`.
         */
        fun parseVarsFromText(content: String?): Map<String, String> {
            val result: MutableMap<String, String> = LinkedHashMap()
            if (content == null || !content.contains("=")) {
                return result
            }

            for (pair in splitOnSeparators(content)) {
                val pos = indexOfUnescaped(pair, '=')
                // no separator, or nothing before it: there is no pair to add
                if (pos <= 0) continue
                result[unescape(pair.substring(0, pos)).trim()] = unescape(pair.substring(pos + 1))
            }
            return result
        }

        /**
         * Renders variables for the clipboard. Unlike the text field writer, this escapes '=' too,
         * so a name containing one survives the round trip.
         */
        internal fun stringifyForCopy(vars: List<Variable>): String {
            val buf = StringBuilder()
            for (variable in vars) {
                if (buf.isNotEmpty()) {
                    buf.append(';')
                }
                buf.append(escape(variable.name, '=', ';'))
                    .append('=')
                    .append(escape(variable.value, '=', ';'))
            }
            return buf.toString()
        }

        /** Escapes [chars] so that [parseVarsFromText] reads the string back unchanged. */
        internal fun escape(s: String, vararg chars: Char): String {
            // escaping anything else would not survive the round trip
            require(chars.all { it in ESCAPABLE }) { "not escapable: ${chars.toList()}" }
            val buf = StringBuilder(s.length)
            for (c in s) {
                // the backslash is always escaped, otherwise the encoding is ambiguous
                if (c == '\\' || chars.contains(c)) {
                    buf.append('\\')
                }
                buf.append(c)
            }
            return buf.toString()
        }

        private fun unescape(s: String): String {
            val buf = StringBuilder(s.length)
            var i = 0
            while (i < s.length) {
                val c = s[i]
                if (c == '\\' && i + 1 < s.length && s[i + 1] in ESCAPABLE) {
                    buf.append(s[i + 1])
                    i += 2
                } else {
                    buf.append(c)
                    i++
                }
            }
            return buf.toString()
        }

        /** Index of the first [target] not preceded by a backslash, or -1. */
        private fun indexOfUnescaped(s: String, target: Char, from: Int = 0): Int {
            var i = from
            while (i < s.length) {
                val c = s[i]
                if (c == '\\' && i + 1 < s.length && s[i + 1] in ESCAPABLE) {
                    i += 2
                    continue
                }
                if (c == target) {
                    return i
                }
                i++
            }
            return -1
        }

        /** Splits on ';', leaving an escaped one inside the value it belongs to. */
        private fun splitOnSeparators(content: String): List<String> {
            val parts = mutableListOf<String>()
            var start = 0
            while (true) {
                val end = indexOfUnescaped(content, ';', start)
                if (end < 0) {
                    parts.add(content.substring(start))
                    return parts
                }
                parts.add(content.substring(start, end))
                start = end + 1
            }
        }
    }
}
